nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-standard-record-utils/src/main/java/org/apache/nifi/schema/validation/StandardSchemaValidator.java

```diff
@@ -233,41 +247,57 @@ private boolean isTypeCorrect(final Object value, final DataType dataType) {
                 }
 
                 return false;
-            case BIGINT:
-                return value instanceof BigInteger;
             case BOOLEAN:
                 return value instanceof Boolean;
-            case BYTE:
-                return value instanceof Byte;
             case CHAR:
                 return value instanceof Character;
             case DATE:
                 return value instanceof java.sql.Date;
-            case DOUBLE:
-                return value instanceof Double;
-            case FLOAT:
-                // Some readers do not provide float vs. double.
-                // We should consider if it makes sense to allow either a Float or a Double here or have
-                // a Reader indicate whether or not it supports higher precision, etc.
-                // Same goes for Short/Integer
-                return value instanceof Float;
-            case INT:
-                return value instanceof Integer;
-            case LONG:
-                return value instanceof Long;
-            case SHORT:
-                return value instanceof Short;
             case STRING:
                 return value instanceof String;
             case TIME:
                 return value instanceof java.sql.Time;
             case TIMESTAMP:
                 return value instanceof java.sql.Timestamp;
+
+            // Numeric data types
+            case BIGINT:
+            case LONG:
+            case INT:
+            case SHORT:
+            case BYTE:
+            case DOUBLE:
+                return isFittingNumberType(value, dataType.getFieldType());
+            case FLOAT:
+                // Some readers do not provide float vs. double.
+                // We should consider if it makes sense to allow either a Float or a Double here or have
+                // a Reader indicate whether or not it supports higher precision, etc.
+                // Same goes for Short/Integer
+                return isFittingNumberType(value, dataType.getFieldType()) || isDoubleWithinFloatInterval(value);
         }
 
         return false;
     }
 
+    /**
+     * Checks if an incoming value satisfies the requirements of a given (numeric) type or any of it's narrow data type.
+     *
+     * @param value Incoming value.
+     * @param fieldType The expected field type.
+     *
+     * @return Returns true if the incoming value satisfies the data type of any of it's narrow data types. Otherwise returns false. Only numeric data types are supported.
+     */
+    private boolean isFittingNumberType(final Object value, final RecordFieldType fieldType) {
+        return NUMERIC_VALIDATORS.containsKey(fieldType)
+            && (NUMERIC_VALIDATORS.get(fieldType).test(value)
+                || fieldType.getNarrowDataTypes().stream().map(type -> NUMERIC_VALIDATORS.get(type)).anyMatch(validator -> validator.test(value))
```

Comment:
I have a feeling that users would expect `DOUBLE` and `FLOAT` to accept integers as well. The current `narrowDataTypes` for those wouldn't allow it.

Reply:
I would agree with that. Especially for something like CSV, you may well have a `Double` field, and that field across 5 lines may be something like `28.38`, `10.9`, `10`, `15.4`, `18`. We should  not consider the `10` and the `18` to be invalid

Reply:
It makes sense if I touch this part, then this should be aimed as well. I will update my proposal with these ideas, thanks!